### ISSUE TITLE
"Error: Request failed with status code 400" with some playlists using getSetInfo()

### ISSUE THREAD

**Issue by supergrov3r** [zackradisic/node-soundcloud-downloader#26]
Hi all - I'm getting a http 400/Bad Request error when using getSetInfo() for some playlists (e.g. https://soundcloud.com/ilyanaazman/sets/best-of-mrrevillz or https://soundcloud.com/ilyanaazman/sets/bestofchillnation) while most others are perfectly fine (e.g. https://soundcloud.com/ilyanaazman/sets/latenightlofi).

Console output below with masked out client_id:

```
Error: Request failed with status code 400      at createError (/workspace/node_modules/axios/lib/core/createError.js:16:15)      at settle (/workspace/node_modules/axios/lib/core/settle.js:17:12)      at IncomingMessage.handleStreamEnd (/workspace/node_modules/axios/lib/adapters/http.js:244:11)      at IncomingMessage.emit (events.js:327:22)      at endReadableNT (_stream_readable.js:1220:12)      at processTicksAndRejections (internal/process/task_queues.js:84:21) {
  config: {
    url: 'https://api-v2.soundcloud.com/tracks?ids=867665638%2C916488845%2C916541744%2C879502579%2C911544727%2C920224540%2C920227672%2C912258409%2C182859485%2C879588352%2C871408492%2C903271243%2C899338447%2C907746022%2C906903244%2C851514058%2C907746013%2C899541523%2C871778755%2C845983675%2C727730749%2C310455689%2C351996389%2C221308049%2C617744415%2C222335592%2C220021334%2C625572180%2C586430472%2C136515969%2C350544130%2C292528042%2C672835793%2C233271745%2C634818096%2C345014116%2C194674830%2C613079007%2C219240973%2C250830766%2C157946069%2C242071122%2C220940419%2C356904608%2C447921720%2C325876880%2C303520330%2C577977645%2C258583921%2C329292478%2C302734389%2C288394390%2C304744944%2C213828110%2C158384606%2C322256287%2C654816296%2C321085206%2C180526491%2C151836932%2C240510101%2C259932466%2C340422811%2C120417156%2C308158931%2C242970980%2C265004302%2C225515495%2C94470145%2C147235179%2C193641840%2C192956375%2C243086787%2C418406040%2C319849669%2C361549280%2C348894700%2C357040427%2C482437833%2C252638815%2C465024060%2C465229422%2C450607776%2C329511908%2C493285392%2C489363333%2C502948368%2C509000139%2C411189549%2C290210997%2C519271257%2C519206883%2C604903338%2C578505426%2C560864886%2C556088364%2C546188349%2C516324954%2C145367831%2C323733248%2C135322127%2C353880398%2C396944928%2C296968053%2C494976372%2C215558123%2C190445096%2C192408098%2C242672575%2C204778816%2C319887152%2C403960782%2C194510401%2C214816945%2C383884892%2C341315415%2C172595409%2C210014181%2C519749160%2C226139958%2C483679920%2C84118776%2C473381727%2C214140973%2C489177243%2C266963313%2C285262888%2C272020529&client_id=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX',
    method: 'get',
    headers: {
      Accept: 'application/json, text/plain, */*',
      'User-Agent': 'axios/0.20.0'
    },
    transformRequest: [ [Function: transformRequest] ],
    transformResponse: [ [Function: transformResponse] ],
    timeout: 0,
    adapter: [Function: httpAdapter],
    xsrfCookieName: 'XSRF-TOKEN',
    xsrfHeaderName: 'X-XSRF-TOKEN',
    maxContentLength: -1,
    maxBodyLength: -1,
    validateStatus: [Function: validateStatus],
    data: undefined
  },
  request: ClientRequest {
    _events: [Object: null prototype] {
      socket: [Function],
      abort: [Function],
      aborted: [Function],
      connect: [Function],
      error: [Function],
      timeout: [Function],
      prefinish: [Function: requestOnPrefinish]
    },
    _eventsCount: 7,
    _maxListeners: undefined,
    outputData: [],
    outputSize: 0,
    writable: true,
    _last: true,
    chunkedEncoding: false,
    shouldKeepAlive: false,
    useChunkedEncodingByDefault: false,
    sendDate: false,
    _removedConnection: false,
    _removedContLen: false,
    _removedTE: false,
    _contentLength: 0,
    _hasBody: true,
    _trailer: '',
    finished: true,
    _headerSent: true,
    socket: TLSSocket {
      _tlsOptions: [Object],
      _secureEstablished: true,
      _securePending: false,
      _newSessionPending: false,
      _controlReleased: true,
      secureConnecting: true,
      _SNICallback: null,
      servername: 'api-v2.soundcloud.com',
      alpnProtocol: false,
      authorized: true,
      authorizationError: null,
      encrypted: true,
      _events: [Object: null prototype],
      _eventsCount: 10,
      connecting: false,
      _hadError: false,
      _parent: null,
      _host: 'api-v2.soundcloud.com',
      _readableState: [ReadableState],
      readable: true,
      _maxListeners: undefined,
      _writableState: [WritableState],
      writable: false,
      allowHalfOpen: false,
      _sockname: null,
      _pendingData: null,
      _pendingEncoding: '',
      server: undefined,
      _server: null,
      ssl: [TLSWrap],
      _requestCert: true,
      _rejectUnauthorized: true,
      parser: null,
      _httpMessage: [Circular],
      [Symbol(res)]: [TLSWrap],
      [Symbol(verified)]: true,
      [Symbol(pendingSession)]: null,
      [Symbol(asyncId)]: 210925,
      [Symbol(kHandle)]: [TLSWrap],
      [Symbol(kSetNoDelay)]: false,
      [Symbol(lastWriteQueueSize)]: 0,
      [Symbol(timeout)]: null,
      [Symbol(kBuffer)]: null,
      [Symbol(kBufferCb)]: null,
      [Symbol(kBufferGen)]: null,
      [Symbol(kCapture)]: false,
      [Symbol(kBytesRead)]: 0,
      [Symbol(kBytesWritten)]: 0,
      [Symbol(connect-options)]: [Object]
    },
    connection: TLSSocket {
      _tlsOptions: [Object],
      _secureEstablished: true,
      _securePending: false,
      _newSessionPending: false,
      _controlReleased: true,
      secureConnecting: true,
      _SNICallback: null,
      servername: 'api-v2.soundcloud.com',
      alpnProtocol: false,
      authorized: true,
      authorizationError: null,
      encrypted: true,
      _events: [Object: null prototype],
      _eventsCount: 10,
      connecting: false,
      _hadError: false,
      _parent: null,
      _host: 'api-v2.soundcloud.com',
      _readableState: [ReadableState],
      readable: true,
      _maxListeners: undefined,
      _writableState: [WritableState],
      writable: false,
      allowHalfOpen: false,
      _sockname: null,
      _pendingData: null,
      _pendingEncoding: '',
      server: undefined,
      _server: null,
      ssl: [TLSWrap],
      _requestCert: true,
      _rejectUnauthorized: true,
      parser: null,
      _httpMessage: [Circular],
      [Symbol(res)]: [TLSWrap],
      [Symbol(verified)]: true,
      [Symbol(pendingSession)]: null,
      [Symbol(asyncId)]: 210925,
      [Symbol(kHandle)]: [TLSWrap],
      [Symbol(kSetNoDelay)]: false,
      [Symbol(lastWriteQueueSize)]: 0,
      [Symbol(timeout)]: null,
      [Symbol(kBuffer)]: null,
      [Symbol(kBufferCb)]: null,
      [Symbol(kBufferGen)]: null,
      [Symbol(kCapture)]: false,
      [Symbol(kBytesRead)]: 0,
      [Symbol(kBytesWritten)]: 0,
      [Symbol(connect-options)]: [Object]
    },
    _header: 'GET /tracks?ids=867665638%2C916488845%2C916541744%2C879502579%2C911544727%2C920224540%2C920227672%2C912258409%2C182859485%2C879588352%2C871408492%2C903271243%2C899338447%2C907746022%2C906903244%2C851514058%2C907746013%2C899541523%2C871778755%2C845983675%2C727730749%2C310455689%2C351996389%2C221308049%2C617744415%2C222335592%2C220021334%2C625572180%2C586430472%2C136515969%2C350544130%2C292528042%2C672835793%2C233271745%2C634818096%2C345014116%2C194674830%2C613079007%2C219240973%2C250830766%2C157946069%2C242071122%2C220940419%2C356904608%2C447921720%2C325876880%2C303520330%2C577977645%2C258583921%2C329292478%2C302734389%2C288394390%2C304744944%2C213828110%2C158384606%2C322256287%2C654816296%2C321085206%2C180526491%2C151836932%2C240510101%2C259932466%2C340422811%2C120417156%2C308158931%2C242970980%2C265004302%2C225515495%2C94470145%2C147235179%2C193641840%2C192956375%2C243086787%2C418406040%2C319849669%2C361549280%2C348894700%2C357040427%2C482437833%2C252638815%2C465024060%2C465229422%2C450607776%2C329511908%2C493285392%2C489363333%2C502948368%2C509000139%2C411189549%2C290210997%2C519271257%2C519206883%2C604903338%2C578505426%2C560864886%2C556088364%2C546188349%2C516324954%2C145367831%2C323733248%2C135322127%2C353880398%2C396944928%2C296968053%2C494976372%2C215558123%2C190445096%2C192408098%2C242672575%2C204778816%2C319887152%2C403960782%2C194510401%2C214816945%2C383884892%2C341315415%2C172595409%2C210014181%2C519749160%2C226139958%2C483679920%2C84118776%2C473381727%2C214140973%2C489177243%2C266963313%2C285262888%2C272020529&client_id=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX HTTP/1.1\r\n' +
      'Accept: application/json, text/plain, */*\r\n' +
      'User-Agent: axios/0.20.0\r\n' +
      'Host: api-v2.soundcloud.com\r\n' +
      'Connection: close\r\n' +
      '\r\n',
    _onPendingData: [Function: noopPendingOutput],
    agent: Agent {
      _events: [Object: null prototype],
      _eventsCount: 2,
      _maxListeners: undefined,
      defaultPort: 443,
      protocol: 'https:',
      options: [Object],
      requests: {},
      sockets: [Object],
      freeSockets: {},
      keepAliveMsecs: 1000,
      keepAlive: false,
      maxSockets: Infinity,
      maxFreeSockets: 256,
      maxCachedSessions: 100,
      _sessionCache: [Object],
      [Symbol(kCapture)]: false
    },
    socketPath: undefined,
    method: 'GET',
    insecureHTTPParser: undefined,
    path: '/tracks?ids=867665638%2C916488845%2C916541744%2C879502579%2C911544727%2C920224540%2C920227672%2C912258409%2C182859485%2C879588352%2C871408492%2C903271243%2C899338447%2C907746022%2C906903244%2C851514058%2C907746013%2C899541523%2C871778755%2C845983675%2C727730749%2C310455689%2C351996389%2C221308049%2C617744415%2C222335592%2C220021334%2C625572180%2C586430472%2C136515969%2C350544130%2C292528042%2C672835793%2C233271745%2C634818096%2C345014116%2C194674830%2C613079007%2C219240973%2C250830766%2C157946069%2C242071122%2C220940419%2C356904608%2C447921720%2C325876880%2C303520330%2C577977645%2C258583921%2C329292478%2C302734389%2C288394390%2C304744944%2C213828110%2C158384606%2C322256287%2C654816296%2C321085206%2C180526491%2C151836932%2C240510101%2C259932466%2C340422811%2C120417156%2C308158931%2C242970980%2C265004302%2C225515495%2C94470145%2C147235179%2C193641840%2C192956375%2C243086787%2C418406040%2C319849669%2C361549280%2C348894700%2C357040427%2C482437833%2C252638815%2C465024060%2C465229422%2C450607776%2C329511908%2C493285392%2C489363333%2C502948368%2C509000139%2C411189549%2C290210997%2C519271257%2C519206883%2C604903338%2C578505426%2C560864886%2C556088364%2C546188349%2C516324954%2C145367831%2C323733248%2C135322127%2C353880398%2C396944928%2C296968053%2C494976372%2C215558123%2C190445096%2C192408098%2C242672575%2C204778816%2C319887152%2C403960782%2C194510401%2C214816945%2C383884892%2C341315415%2C172595409%2C210014181%2C519749160%2C226139958%2C483679920%2C84118776%2C473381727%2C214140973%2C489177243%2C266963313%2C285262888%2C272020529&client_id=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX',
    _ended: true,
    res: IncomingMessage {
      _readableState: [ReadableState],
      readable: false,
      _events: [Object: null prototype],
      _eventsCount: 3,
      _maxListeners: undefined,
      socket: [TLSSocket],
      connection: [TLSSocket],
      httpVersionMajor: 1,
      httpVersionMinor: 1,
      httpVersion: '1.1',
      complete: true,
      headers: [Object],
      rawHeaders: [Array],
      trailers: {},
      rawTrailers: [],
      aborted: false,
      upgrade: false,
      url: '',
      method: null,
      statusCode: 400,
      statusMessage: 'Bad Request',
      client: [TLSSocket],
      _consuming: false,
      _dumped: false,
      req: [Circular],
      responseUrl: 'https://api-v2.soundcloud.com/tracks?ids=867665638%2C916488845%2C916541744%2C879502579%2C911544727%2C920224540%2C920227672%2C912258409%2C182859485%2C879588352%2C871408492%2C903271243%2C899338447%2C907746022%2C906903244%2C851514058%2C907746013%2C899541523%2C871778755%2C845983675%2C727730749%2C310455689%2C351996389%2C221308049%2C617744415%2C222335592%2C220021334%2C625572180%2C586430472%2C136515969%2C350544130%2C292528042%2C672835793%2C233271745%2C634818096%2C345014116%2C194674830%2C613079007%2C219240973%2C250830766%2C157946069%2C242071122%2C220940419%2C356904608%2C447921720%2C325876880%2C303520330%2C577977645%2C258583921%2C329292478%2C302734389%2C288394390%2C304744944%2C213828110%2C158384606%2C322256287%2C654816296%2C321085206%2C180526491%2C151836932%2C240510101%2C259932466%2C340422811%2C120417156%2C308158931%2C242970980%2C265004302%2C225515495%2C94470145%2C147235179%2C193641840%2C192956375%2C243086787%2C418406040%2C319849669%2C361549280%2C348894700%2C357040427%2C482437833%2C252638815%2C465024060%2C465229422%2C450607776%2C329511908%2C493285392%2C489363333%2C502948368%2C509000139%2C411189549%2C290210997%2C519271257%2C519206883%2C604903338%2C578505426%2C560864886%2C556088364%2C546188349%2C516324954%2C145367831%2C323733248%2C135322127%2C353880398%2C396944928%2C296968053%2C494976372%2C215558123%2C190445096%2C192408098%2C242672575%2C204778816%2C319887152%2C403960782%2C194510401%2C214816945%2C383884892%2C341315415%2C172595409%2C210014181%2C519749160%2C226139958%2C483679920%2C84118776%2C473381727%2C214140973%2C489177243%2C266963313%2C285262888%2C272020529&client_id=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX',
      redirects: [],
      [Symbol(kCapture)]: false
    },
    aborted: false,
    timeoutCb: null,
    upgradeOrConnect: false,
    parser: null,
    maxHeadersCount: null,
    reusedSocket: false,
    _redirectable: Writable {
      _writableState: [WritableState],
      writable: true,
      _events: [Object: null prototype],
      _eventsCount: 2,
      _maxListeners: undefined,
      _options: [Object],
      _ended: true,
      _ending: true,
      _redirectCount: 0,
      _redirects: [],
      _requestBodyBuffers: [],
      _onNativeResponse: [Function],
      _currentRequest: [Circular],
      _currentUrl: 'https://api-v2.soundcloud.com/tracks?ids=867665638%2C916488845%2C916541744%2C879502579%2C911544727%2C920224540%2C920227672%2C912258409%2C182859485%2C879588352%2C871408492%2C903271243%2C899338447%2C907746022%2C906903244%2C851514058%2C907746013%2C899541523%2C871778755%2C845983675%2C727730749%2C310455689%2C351996389%2C221308049%2C617744415%2C222335592%2C220021334%2C625572180%2C586430472%2C136515969%2C350544130%2C292528042%2C672835793%2C233271745%2C634818096%2C345014116%2C194674830%2C613079007%2C219240973%2C250830766%2C157946069%2C242071122%2C220940419%2C356904608%2C447921720%2C325876880%2C303520330%2C577977645%2C258583921%2C329292478%2C302734389%2C288394390%2C304744944%2C213828110%2C158384606%2C322256287%2C654816296%2C321085206%2C180526491%2C151836932%2C240510101%2C259932466%2C340422811%2C120417156%2C308158931%2C242970980%2C265004302%2C225515495%2C94470145%2C147235179%2C193641840%2C192956375%2C243086787%2C418406040%2C319849669%2C361549280%2C348894700%2C357040427%2C482437833%2C252638815%2C465024060%2C465229422%2C450607776%2C329511908%2C493285392%2C489363333%2C502948368%2C509000139%2C411189549%2C290210997%2C519271257%2C519206883%2C604903338%2C578505426%2C560864886%2C556088364%2C546188349%2C516324954%2C145367831%2C323733248%2C135322127%2C353880398%2C396944928%2C296968053%2C494976372%2C215558123%2C190445096%2C192408098%2C242672575%2C204778816%2C319887152%2C403960782%2C194510401%2C214816945%2C383884892%2C341315415%2C172595409%2C210014181%2C519749160%2C226139958%2C483679920%2C84118776%2C473381727%2C214140973%2C489177243%2C266963313%2C285262888%2C272020529&client_id=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX',
      [Symbol(kCapture)]: false
    },
    [Symbol(kCapture)]: false,
    [Symbol(kNeedDrain)]: false,
    [Symbol(corked)]: 0,
    [Symbol(kOutHeaders)]: [Object: null prototype] {
      accept: [Array],
      'user-agent': [Array],
      host: [Array]
    }
  },
  response: {
    status: 400,
    statusText: 'Bad Request',
    headers: {
      'strict-transport-security': 'max-age=63072000',
      vary: 'Origin',
      via: '1.1 e57fe70b9ed429fb51b4b2432cadc67b.cloudfront.net (CloudFront)',
      'x-amz-cf-id': 'YbS6yqSmjIrGDdLZqKdTd3QHiqETWVPsKNUV43eE_Zak-hrQqsoAAQ=='
    config: {
      url: 'https://api-v2.soundcloud.com/tracks?ids=867665638%2C916488845%2C916541744%2C879502579%2C911544727%2C920224540%2C920227672%2C912258409%2C182859485%2C879588352%2C871408492%2C903271243%2C899338447%2C907746022%2C906903244%2C851514058%2C907746013%2C899541523%2C871778755%2C845983675%2C727730749%2C310455689%2C351996389%2C221308049%2C617744415%2C222335592%2C220021334%2C625572180%2C586430472%2C136515969%2C350544130%2C292528042%2C672835793%2C233271745%2C634818096%2C345014116%2C194674830%2C613079007%2C219240973%2C250830766%2C157946069%2C242071122%2C220940419%2C356904608%2C447921720%2C325876880%2C303520330%2C577977645%2C258583921%2C329292478%2C302734389%2C288394390%2C304744944%2C213828110%2C158384606%2C322256287%2C654816296%2C321085206%2C180526491%2C151836932%2C240510101%2C259932466%2C340422811%2C120417156%2C308158931%2C242970980%2C265004302%2C225515495%2C94470145%2C147235179%2C193641840%2C192956375%2C243086787%2C418406040%2C319849669%2C361549280%2C348894700%2C357040427%2C482437833%2C252638815%2C465024060%2C465229422%2C450607776%2C329511908%2C493285392%2C489363333%2C502948368%2C509000139%2C411189549%2C290210997%2C519271257%2C519206883%2C604903338%2C578505426%2C560864886%2C556088364%2C546188349%2C516324954%2C145367831%2C323733248%2C135322127%2C353880398%2C396944928%2C296968053%2C494976372%2C215558123%2C190445096%2C192408098%2C242672575%2C204778816%2C319887152%2C403960782%2C194510401%2C214816945%2C383884892%2C341315415%2C172595409%2C210014181%2C519749160%2C226139958%2C483679920%2C84118776%2C473381727%2C214140973%2C489177243%2C266963313%2C285262888%2C272020529&client_id=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX',
      method: 'get',
      headers: [Object],
      transformRequest: [Array],
      transformResponse: [Array],
      timeout: 0,
      adapter: [Function: httpAdapter],
      xsrfHeaderName: 'X-XSRF-TOKEN',
      maxBodyLength: -1,
    },
      _events: [Object: null prototype],
      _maxListeners: undefined,
      writable: true,
      chunkedEncoding: false,
      useChunkedEncodingByDefault: false,
      _removedConnection: false,
      _contentLength: 0,
      _trailer: '',
      _headerSent: true,
      connection: [TLSSocket],
        'Accept: application/json, text/plain, */*\r\n' +
        'User-Agent: axios/0.20.0\r\n' +
        '\r\n',
      agent: [Agent],
      socketPath: undefined,
      insecureHTTPParser: undefined,
      _ended: true,
      timeoutCb: null,
      [Symbol(kNeedDrain)]: false,
      [Symbol(kOutHeaders)]: [Object: null prototype]
    data: {}
  isAxiosError: true,
}

```

**Comment by zackradisic** [comment]
Thanks for bringing this up, I'll take a look at it later when I get the chance. Given that the first two playlists you've linked have 100+ tracks and the third playlist — which works — only has 48, my speculation is that the problem is the number of tracks in the playlists. 

The SoundCloud API endpoint for retrieving track information might have a limit on how many tracks can be queried at once, hopefully a simple solution like distributing the tracks among multiple API requests will work.

**Comment by supergrov3r** [comment]
Thanks for taking a look at it.

I think you might be on to something. I created a new playlist and ran into issues when the number of tracks > 55.

Hopefully that helps.

**Comment by zackradisic** [comment]
Just confirmed that the SoundCloud API endpoint used to retrieve track information in `getSetInfo()` is limited to only 50 tracks at a time. Info for the first 5 is supplied by the initial request, so it makes sense that this function breaks when the track count of a playlist is 56 or greater.

I altered `getSetInfo()` to make multiple requests of maximum size 50 tracks if the track length of the playlist is greater than 55.

**Comment by zackradisic** [comment]
@supergrov3r Just published [v0.1.9](https://github.com/zackradisic/node-soundcloud-downloader/releases/tag/v0.1.9) which includes a fix for this problem.

**Comment by supergrov3r** [comment]
Confirmed v0.1.9 fixes this - thanks again!
